### PR TITLE
[CDAP-16353] Add scaling factor to adjust resource size for KubeTwill program

### PR DESCRIPTION
**Pull request by hsiaoching**

### CODE REVIEW

**Comment by hsiaoching** [comment]
@googlebot I fixed it.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java [hunk 1]
```diff
@@ -96,14 +96,16 @@
   private final Map<String, Map<String, String>> environments;
   private final RunId twillRunId;
   private final V1ObjectMeta resourceMeta;
+  private final Map<String, Float> kubeScalingFactors;
```
Comment (wyzhang) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java]: How about adding "resource" to the name, e.g. kubeResourceScalingFactors

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 2]
```diff
@@ -74,6 +74,8 @@
   private static final String INSTANCE_LABEL = "master.environment.k8s.instance.label";
   // Label for the container name
   private static final String CONTAINER_LABEL = "master.environment.k8s.container.label";
+  private static final String CPU_SCALING_FACTOR = "master.environment.k8s.container.cpu.scaling.factor";
```
Comment (wyzhang) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: nit: maybe just shorten the name to CPU_MULTIPLIER

Comment (wyzhang) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: same for other places scaling_factor -> multiplier

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java (hunk 1 above)
Comment (wyzhang) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java]: Would it be simpler to keep a pointer to conf in KubeTwillRunnerService and KubeTwillPreparer?

Then you would need to add const string for the keys 
kubernetes.request.memory.multiplier and kubernetes.request.cpu.multiplier to 
io.cdap.cdap.common.conf.Constants

Set the default value in 
cdap-common/src/main/resources/cdap-default.xml

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java [hunk 3]
```diff
@@ -149,7 +152,12 @@ public void initialize(MasterEnvironmentContext context) throws IOException, Api
                namespace, podKillerSelector, delayMillis);
     }
 
-    twillRunner = new KubeTwillRunnerService(namespace, discoveryService, podInfo, resourcePrefix,
+    Map<String, Float> kubeScalingFactors = new HashMap<String, Float>();
+    kubeScalingFactors.put("cpu", Float.parseFloat(conf.getOrDefault(CPU_SCALING_FACTOR, DEFAULT_SCALING_FACTOR)));
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: Why need to copy the config? You can pass the whole `conf` object to the runner service. In future there could be other configurations being added.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java [hunk 4]
```diff
@@ -453,9 +456,10 @@ V1Deployment buildDeployment(RunId runid,
 
     V1ResourceRequirements resourceRequirements = new V1ResourceRequirements();
     Map<String, Quantity> quantityMap = new HashMap<>();
-    quantityMap.put("cpu", new Quantity(String.valueOf(vCores)));
+    quantityMap.put("cpu", new Quantity(String.valueOf((float) (vCores * kubeScalingFactors.get("cpu")))));
```
Comment (chtyim) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java]: Does `Quantity` takes floating point value? Should we round it up/down?

Comment (hsiaoching) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java]: Quantity converts the string number to BigDecimal so can be float. Ref: https://github.com/kubernetes-client/java/blob/master/kubernetes/src/main/java/io/kubernetes/client/custom/Quantity.java#L17

The default value of vCores is 1. Rounding up or down would remove the scaling effect we want here. Ref: https://github.com/apache/twill/blob/master/twill-api/src/main/java/org/apache/twill/api/ResourceSpecification.java#L28

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java [hunk 5]
```diff
@@ -96,14 +99,16 @@
   private final Map<String, Map<String, String>> environments;
   private final RunId twillRunId;
   private final V1ObjectMeta resourceMeta;
+  private final Map<String, String> kubeMasterConf;
```
Comment (wyzhang) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java]: nitpicking: maybe name it conf

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java [hunk 6]
```diff
@@ -453,9 +459,11 @@ V1Deployment buildDeployment(RunId runid,
 
     V1ResourceRequirements resourceRequirements = new V1ResourceRequirements();
     Map<String, Quantity> quantityMap = new HashMap<>();
-    quantityMap.put("cpu", new Quantity(String.valueOf(vCores)));
+    float cpuMultiplier = Float.parseFloat(kubeMasterConf.getOrDefault(CPU_MULTIPLIER, DEFAULT_MULTIPLIER));
+    float memoryMultiplier = Float.parseFloat(kubeMasterConf.getOrDefault(MEMORY_MULTIPLIER, DEFAULT_MULTIPLIER));
+    quantityMap.put("cpu", new Quantity(String.valueOf((float) (vCores * cpuMultiplier))));
```
Comment (wyzhang) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java]: how about using "milli" unit for CPU with rounding up like this?

quantityMap.put("cpu", new Quantity(String.format("%dm", (int)Math.ceil((float)vCores * 1000 * multiplier))

You can create a util function 
vCoresToMilliCpu(vcore, multplier) {
}

**Review by wyzhang — APPROVED** [review]

File: cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java (hunk 3 above)
Comment (hsiaoching) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java]: Done.

**Pull request merged**